### ISSUE TITLE
Gauge Metrics for Karpenter NodeClaims

### ISSUE THREAD

**Issue by msvechla** [kubernetes-sigs/karpenter#1276]
### Description

**What problem are you trying to solve?**

When operating clusters it is helpful to know how many `NodeClaims` currently have drifted from the desired state, as we eventually want to get back to zero drift when a configuration change was rolled out.

We already have a counter metric for these events, like `karpenter_nodeclaims_drifted`, however knowing the current exact number of drifted `NodeClaims` would be better tracked with a gauge.

**How important is this feature to you?**

I would be open to try to contribute via PR for this

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by sftim** [comment]
Can you get the number you want using a recording rule or math? If you can, let's recommend that (and avoid emitting an extra time series).

**Comment by sftim** [comment]
If you can't get that number, is there another counter time series we can emit that would allow it?

**Comment by msvechla** [comment]
To my knowledge using a counter will not work, as a counter can only increase. So for me only a gauge metric makes sense in general for the nodeclaim metrics. With the current counter metrics you can just see if the value increased, not if it decreased, or what the current actual value is.

It might be useful for detecting spikes in specific events, but its not helpful in determining the current state.

See also https://prometheus.io/docs/tutorials/understanding_metric_types/#counter vs https://prometheus.io/docs/tutorials/understanding_metric_types/#gauge

**Comment by sftim** [comment]
> To my knowledge using a counter will not work, as a counter can only increase.

You're right that one counter doesn't work @msvechla. One way to address this is use two (or more) counters.

For example: `karpenter_nodeclaims_drifted_total` and `karpenter_nodeclaims_drift_cleared`. The second one is incremented when a NodeClaim that has drifted is either deleted or somehow fixed. Subtract cleared from the total and you should have the current drifted count, or you can use a recording rule.

I think it's useful to be able to observe when drift is (or isn't!) getting cleared, which isn't as easy to spot from `karpenter_nodeclaims_drifted` plus a gauge. Even so, I can see how a gauge is useful too.

---

Another counter that can be useful to record, per-NodePool, is the (counter) sum of time that NodeClaims have been drifted. For example if you have 10 NodeClaims and 9 of them drift at the same moment, then 60 seconds later the counter will be 60 × 9 higher. You could alarm on seeing this increase by more than _n_ per second (implying that you have drifted NodeClaims not being fixed).

**Comment by msvechla** [comment]
In theory two or more counters work, but from my experience this is unreliable in practice. Counters are not designed to track exact values, so it can happen that a scrape misses a counter increase and then subtracting the values will no longer become 0, even if all drifts have been cleared.

E.g. a counter increases to 5 before it gets scraped the first time, now after the initial scrape, the initial increases are missing.

So its usually impossible to get exact values with counter math. If we want to create alerts based on those metrics (e.g. at least one drifting nodeclaim for X amount of time), being exact is important to avoid false positives.

I'm open to the idea if someone can provide a good example, but from my current experience this always creates issues, while gauges simply work for this scenario

**Comment by msvechla** [comment]
Is there an update on this, or are merge requests accepted? We are often having the issue that quite a few drifted nodes remain in a cluster, and there is currently no way for us to detect this.

This is critical if we are rolling out configuration changes to our nodes and we have no awareness via metrics, whether all nodes have been replaced and are using the new setup

**Comment by jonathan-innis** [comment]
This should be available now! 🎉  We have `operator_nodeclaim_status_condition...` prefixed metrics, including `operator_nodeclaim_status_condition_count` which should give you the number of NodeClaims that are in a Drfited state. 

/close
